experimental/metal/direct_command_buffer.m

```diff
@@ -0,0 +1,1187 @@
+          computeCommandEncoderWithDispatchType:command_buffer->dispatch_type] retain];  // +1
+    }
+  }
+
+  return command_buffer->state.compute_encoder;
+}
+
+static id<MTLBlitCommandEncoder> iree_hal_metal_get_or_begin_blit_encoder(
+    iree_hal_metal_command_buffer_t* command_buffer) {
+  id<MTLCommandBuffer> metal_handle = command_buffer->command_buffer;
+
+  // If we are switching encoders, we would need to use a fence to synchronize "one or more
+  // resources across different passes within a command buffer."
+  // https://developer.apple.com/documentation/metal/resource_synchronization
+  uint64_t encoder_event_value = 0;
+  if (command_buffer->state.compute_encoder) {
+    iree_hal_metal_end_compute_encoder(command_buffer);
+    encoder_event_value = command_buffer->state.next_encoder_event_value++;
+    [metal_handle encodeSignalEvent:command_buffer->state.encoder_event value:encoder_event_value];
+  }
+
+  if (!command_buffer->state.blit_encoder) {
+    if (encoder_event_value != 0) {
+      [metal_handle encodeWaitForEvent:command_buffer->state.encoder_event
+                                 value:encoder_event_value];
+    }
+    @autoreleasepool {  // Use @autoreleasepool to trigger the autorelease within encoder creation.
+      command_buffer->state.blit_encoder = [[metal_handle blitCommandEncoder] retain];  // +1
+    }
+  }
+
+  return command_buffer->state.blit_encoder;
+}
+
+iree_status_t iree_hal_metal_direct_command_buffer_create(
+    iree_hal_device_t* device, iree_hal_command_buffer_mode_t mode,
+    iree_hal_command_category_t command_categories, iree_host_size_t binding_capacity,
+    iree_hal_metal_command_buffer_resource_reference_mode_t resource_reference_mode,
+    id<MTLCommandQueue> queue, iree_allocator_t host_allocator, iree_arena_block_pool_t* block_pool,
+    iree_hal_metal_staging_buffer_t* staging_buffer,
+    iree_hal_metal_builtin_executable_t* builtin_executable,
+    iree_hal_command_buffer_t** out_command_buffer) {
+  IREE_ASSERT_ARGUMENT(device);
+  IREE_ASSERT_ARGUMENT(out_command_buffer);
+  IREE_ASSERT_TRUE(iree_all_bits_set(mode, IREE_HAL_COMMAND_BUFFER_MODE_ONE_SHOT));
+  IREE_ASSERT_TRUE(!iree_any_bit_set(mode, IREE_HAL_COMMAND_BUFFER_MODE_NESTED));
+  *out_command_buffer = NULL;
+
+  if (binding_capacity > 0) {
+    // TODO(#10144): support indirect command buffers with binding tables.
+    return iree_make_status(IREE_STATUS_UNIMPLEMENTED, "unimplemented indirect command buffers");
+  }
+
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_metal_command_buffer_t* command_buffer = NULL;
+  iree_status_t status =
+      iree_allocator_malloc(host_allocator, sizeof(*command_buffer), (void**)&command_buffer);
+  if (iree_status_is_ok(status)) {
+    iree_hal_command_buffer_initialize(
+        device, mode, command_categories, IREE_HAL_QUEUE_AFFINITY_ANY, binding_capacity,
+        &iree_hal_metal_command_buffer_vtable, &command_buffer->base);
```

Comment:
I think there's a logic error in here if the malloc fails. The code after the if expects `command_buffer->base` to be set:
```c
  *out_command_buffer = &command_buffer->base;
  IREE_TRACE_ZONE_END(z0);
  return status;
```

Reply:
Good catch. Moved to the previous `if` case.